File: srpc/mux.go
package srpc

import "sync"

// Mux contains a set of <service, method> handlers.
type Mux interface {
	// Invoker invokes the methods.
	Invoker

	// Register registers a new RPC method handler (service).
	Register(handler Handler) error
	// HasService checks if the service ID exists in the handlers.
	HasService(serviceID string) bool
	// HasServiceMethod checks if <service-id, method-id> exists in the handlers.
	HasServiceMethod(serviceID, methodID string) bool
}

// muxMethods is a mapping from method id to handler.
type muxMethods map[string]Handler

// mux is the default implementation of Mux.
type mux struct {
	// fallback is the list of fallback invokers
	// if the mux doesn't match the service, calls the invokers.
	fallback []Invoker
	// rmtx guards below fields
	rmtx sync.RWMutex
	// services contains a mapping from services to handlers.
	services map[string]muxMethods
}

// NewMux constructs a new Mux.
//
// fallbackInvokers is the list of fallback Invokers to call in the case that
// the service/method is not found on this mux.
func NewMux(fallbackInvokers ...Invoker) Mux {
	return &mux{
		fallback: fallbackInvokers,
		services: make(map[string]muxMethods),
	}
}

// Register registers a new RPC method handler (service).
func (m *mux) Register(handler Handler) error {
	serviceID := handler.GetServiceID()
	methodIDs := handler.GetMethodIDs()
	if serviceID == "" {
		return ErrEmptyServiceID
	}

	m.rmtx.Lock()
	defer m.rmtx.Unlock()

	serviceMethods := m.services[serviceID]
	if serviceMethods == nil {
		serviceMethods = make(muxMethods)
		m.services[serviceID] = serviceMethods
	}
	for _, methodID := range methodIDs {
		if methodID != "" {
			serviceMethods[methodID] = handler
		}
	}

	return nil
}

// HasService checks if the service ID exists in the handlers.
func (m *mux) HasService(serviceID string) bool {
	if serviceID == "" {
		return false
	}

	m.rmtx.Lock()
	defer m.rmtx.Unlock()

	return len(m.services[serviceID]) != 0
}

// HasServiceMethod checks if <service-id, method-id> exists in the handlers.
func (m *mux) HasServiceMethod(serviceID, methodID string) bool {
	if serviceID == "" || methodID == "" {
		return false
	}

	m.rmtx.Lock()
	defer m.rmtx.Unlock()

	handlers := m.services[serviceID]
	for _, mh := range handlers {
		for _, mhMethodID := range mh.GetMethodIDs() {
			if mhMethodID == methodID {
				return true
			}
		}
	}

	return false
}

// InvokeMethod invokes the method matching the service & method ID.
// Returns false, nil if not found.
// If service string is empty, ignore it.
func (m *mux) InvokeMethod(serviceID, methodID string, strm Stream) (bool, error) {
	var handler Handler
	m.rmtx.RLock()
	if serviceID == "" {
		for _, svc := range m.services {
			if handler = svc[methodID]; handler != nil {
				break
			}
		}
	} else {
		svcMethods := m.services[serviceID]
		if svcMethods != nil {
			handler = svcMethods[methodID]
		}
	}
	m.rmtx.RUnlock()

	if handler != nil {
		return handler.InvokeMethod(serviceID, methodID, strm)
	}

	for _, invoker := range m.fallback {
		if invoker != nil {
			handled, err := invoker.InvokeMethod(serviceID, methodID, strm)
			if err != nil || handled {
				return handled, err
			}
		}
	}

	return false, nil
}

// _ is a type assertion
var _ Mux = ((*mux)(nil))

File: srpc/mux.cpp
//go:build deps_only

#include "mux.hpp"
#include <mutex>

namespace starpc {

Mux::Mux(std::vector<Invoker*> fallback_invokers)
    : fallback_(std::move(fallback_invokers)) {}

Error Mux::Register(Handler* handler) {
  const std::string& service_id = handler->GetServiceID();
  auto method_ids = handler->GetMethodIDs();

  if (service_id.empty()) {
    return Error::EmptyServiceID;
  }

  std::unique_lock<std::shared_mutex> lock(mtx_);

  auto& service_methods = services_[service_id];
  for (const auto& method_id : method_ids) {
    if (!method_id.empty()) {
      service_methods[method_id] = handler;
    }
  }

  return Error::OK;
}

bool Mux::HasService(const std::string& service_id) const {
  if (service_id.empty()) {
    return false;
  }

  std::shared_lock<std::shared_mutex> lock(mtx_);
  auto it = services_.find(service_id);
  return it != services_.end() && !it->second.empty();
}

bool Mux::HasServiceMethod(const std::string& service_id, const std::string& method_id) const {
  if (service_id.empty() || method_id.empty()) {
    return false;
  }

  std::shared_lock<std::shared_mutex> lock(mtx_);
  auto svc_it = services_.find(service_id);
  if (svc_it == services_.end()) {
    return false;
  }

  for (const auto& [mid, handler] : svc_it->second) {
    for (const auto& handler_method : handler->GetMethodIDs()) {
      if (handler_method == method_id) {
        return true;
      }
    }
  }

  return false;
}

std::pair<bool, Error> Mux::InvokeMethod(
    const std::string& service_id,
    const std::string& method_id,
    Stream* strm) {
  Handler* handler = nullptr;

  {
    std::shared_lock<std::shared_mutex> lock(mtx_);

    if (service_id.empty()) {
      // If service string is empty, search all services
      for (const auto& [svc_id, svc_methods] : services_) {
        auto it = svc_methods.find(method_id);
        if (it != svc_methods.end()) {
          handler = it->second;
          break;
        }
      }
    } else {
      auto svc_it = services_.find(service_id);
      if (svc_it != services_.end()) {
        auto method_it = svc_it->second.find(method_id);
        if (method_it != svc_it->second.end()) {
          handler = method_it->second;
        }
      }
    }
  }

  if (handler != nullptr) {
    return handler->InvokeMethod(service_id, method_id, strm);
  }

  // Try fallback invokers
  for (auto* invoker : fallback_) {
    if (invoker != nullptr) {
      auto [handled, err] = invoker->InvokeMethod(service_id, method_id, strm);
      if (err != Error::OK || handled) {
        return {handled, err};
      }
    }
  }

  return {false, Error::OK};
}

}  // namespace starpc
